File: php/src/AccessPolicy/AccessPolicyEvaluator.php
<?php

declare(strict_types=1);

namespace ProofMark\ShowAd\AccessPolicy;

use Closure;
use ProofMark\ShowAd\Request\RequestContext;

/**
 * Server-only access policy that runs before ShowAd verification.
 *
 * Pipeline (in order):
 *   1. Verified crawler (UA + trusted IP range OR Cloudflare verified-bot OR rDNS)
 *   2. CIDR allowlist resolved from a trusted IP header
 *   3. Publisher-defined `before_protect` callback (premium users, app sessions, ...)
 *
 * UA detection alone is NEVER sufficient to allow access. It only narrows
 * which IP/rDNS rules to apply for the matching family.
 *
 * This is the framework-agnostic counterpart to the Laravel SDK's evaluator.
 * Behaviour is byte-for-byte equivalent on the wire.
 */
final class AccessPolicyEvaluator
{
    /**
     * Default crawler user-agent fragments per family.
     *
     * @var array<string, array<int, string>>
     */
    public const DEFAULT_CRAWLER_USER_AGENTS = [
        'google' => ['googlebot', 'google-inspectiontool', 'apis-google'],
        'bing' => ['bingbot'],
        'duckduckgo' => ['duckduckbot'],
        'yandex' => ['yandexbot'],
        'baidu' => ['baiduspider'],
        'openai' => ['gptbot', 'chatgpt-user', 'oai-searchbot'],
        'anthropic' => ['claudebot', 'anthropic-ai'],
        'perplexity' => ['perplexitybot'],
        'commoncrawl' => ['ccbot'],
        'facebook' => ['facebookexternalhit', 'facebot'],
        'twitter' => ['twitterbot'],
        'linkedin' => ['linkedinbot'],
    ];

    /**
     * Evaluate the access policy for a request.
     *
     * @param array<string, mixed> $config
     * @return array{action: string, reason?: string, redirect_url?: string}
     */
    public function evaluate(RequestContext $request, array $config): array
    {
        $clientIp = $this->resolveClientIp($request, $config['trusted_ip_headers'] ?? []);
        $userAgent = $request->userAgent();

        $crawler = $this->verifyCrawler(
            $clientIp,
            $userAgent,
            is_array($config['crawler'] ?? null) ? $config['crawler'] : [],
            $request
        );

        if ($crawler['verified']) {
            return [
                'action' => 'allow',
                'reason' => 'crawler:' . ($crawler['family'] ?? 'unknown'),
            ];
        }

        $cidrs = isset($config['allow_cidrs']) && is_array($config['allow_cidrs'])
            ? $config['allow_cidrs']
            : [];

        if ($clientIp !== null && $clientIp !== '' && $this->ipInCidrs($clientIp, $cidrs)) {
            return ['action' => 'allow', 'reason' => 'cidr_allowlist'];
        }

        $callback = $config['before_protect'] ?? null;
        if ($callback instanceof Closure) {
            $decision = $callback($request, [
                'client_ip' => $clientIp,
                'user_agent' => $userAgent,
            ]);
            return $this->normaliseDecision($decision);
        }

        return ['action' => 'continue'];
    }

    /**
     * Resolve the client IP from a configured list of trusted edge headers,
     * falling back to the request-level remote address.
     *
     * @param array<int, string> $trustedIpHeaders
     */
    public function resolveClientIp(RequestContext $request, array $trustedIpHeaders): ?string
    {
        foreach ($trustedIpHeaders as $header) {
            $value = $request->header((string) $header);
            if ($value === null || $value === '') {
                continue;
            }
            $first = trim(explode(',', $value)[0]);
            if ($first !== '') {
                return $first;
            }
        }

        $ip = $request->ip;
        return $ip === '' ? null : $ip;
    }

    /**
     * Verify a crawler request by combining a known user-agent family with
     * trusted IP-range or rDNS evidence.
     *
     * @param array<string, mixed> $crawlerConfig
     * @return array{verified: bool, reason: string, family?: string}
     */
    public function verifyCrawler(?string $ip, string $userAgent, array $crawlerConfig, ?RequestContext $request = null): array
    {
        if (empty($crawlerConfig['enabled'])) {
            return ['verified' => false, 'reason' => 'disabled'];
        }

        $families = isset($crawlerConfig['families']) && is_array($crawlerConfig['families'])
            ? $crawlerConfig['families']
            : array_keys(self::DEFAULT_CRAWLER_USER_AGENTS);

        $userAgents = isset($crawlerConfig['user_agents']) && is_array($crawlerConfig['user_agents'])
            ? $crawlerConfig['user_agents']
            : self::DEFAULT_CRAWLER_USER_AGENTS;

        $family = $this->matchCrawlerFamily($userAgent, $families, $userAgents);
        if ($family === null) {
            return ['verified' => false, 'reason' => 'no_family_match'];
        }

        if ($ip === null || $ip === '') {
            return ['verified' => false, 'reason' => 'missing_ip', 'family' => $family];
        }

        if (!empty($crawlerConfig['allow_cloudflare_verified_bot']) && $request !== null) {
            $verifiedBot = $request->header('cf-verified-bot')
                ?? $request->header('x-proofmark-cf-verified-bot');
            if ($verifiedBot !== null && in_array(strtolower($verifiedBot), ['1', 'true', 'yes', 'on'], true)) {
                return ['verified' => true, 'reason' => 'cloudflare_verified_bot', 'family' => $family];
            }
        }

        $cidrs = [];
        if (isset($crawlerConfig['family_cidrs'][$family]) && is_array($crawlerConfig['family_cidrs'][$family])) {
            $cidrs = $crawlerConfig['family_cidrs'][$family];
        }
        if ($this->ipInCidrs($ip, $cidrs)) {
            return ['verified' => true, 'reason' => 'cidr_match', 'family' => $family];
        }

        $verifier = $crawlerConfig['reverse_dns_verifier'] ?? null;
        if ($verifier instanceof Closure && $verifier($ip, $family)) {
            return ['verified' => true, 'reason' => 'reverse_dns_match', 'family' => $family];
        }

        return ['verified' => false, 'reason' => 'ip_not_verified', 'family' => $family];
    }

    /**
     * @param array<int, string> $cidrs
     */
    public function ipInCidrs(string $ip, array $cidrs): bool
    {
        foreach ($cidrs as $cidr) {
            if ($this->ipMatchesCidr($ip, (string) $cidr)) {
                return true;
            }
        }
        return false;
    }

    /**
     * @param array<int, string> $families
     * @param array<string, array<int, string>> $userAgentMap
     */
    private function matchCrawlerFamily(string $userAgent, array $families, array $userAgentMap): ?string
    {
        $needle = strtolower($userAgent);
        if ($needle === '') {
            return null;
        }

        foreach ($families as $family) {
            $family = (string) $family;
            $fragments = $userAgentMap[$family] ?? [];
            foreach ($fragments as $fragment) {
                $fragment = (string) $fragment;
                if ($fragment !== '' && strpos($needle, strtolower($fragment)) !== false) {
                    return $family;
                }
            }
        }
        return null;
    }

    /**
     * @param mixed $decision
     * @return array{action: string, reason?: string, redirect_url?: string}
     */
    private function normaliseDecision($decision): array
    {
        if (is_string($decision)) {
            return ['action' => $decision];
        }
        if (is_array($decision) && isset($decision['action'])) {
            return $decision;
        }
        return ['action' => 'continue'];
    }

    private function ipMatchesCidr(string $ip, string $cidr): bool
    {
        if (strpos($cidr, '/') === false) {
            $a = @inet_pton($ip);
            $b = @inet_pton($cidr);
            return $a !== false && $b !== false && $a === $b;
        }

        [$range, $bits] = explode('/', $cidr, 2);
        if (!is_numeric($bits)) {
            return false;
        }

        $rangeBin = @inet_pton($range);
        $ipBin = @inet_pton($ip);
        if ($rangeBin === false || $ipBin === false || strlen($rangeBin) !== strlen($ipBin)) {
            return false;
        }

        $bits = (int) $bits;
        $maxBits = strlen($ipBin) * 8;
        if ($bits < 0 || $bits > $maxBits) {
            return false;
        }

        $bytes = intdiv($bits, 8);
        $remainder = $bits % 8;

        if ($bytes > 0 && substr($rangeBin, 0, $bytes) !== substr($ipBin, 0, $bytes)) {
            return false;
        }

        if ($remainder === 0) {
            return true;
        }

        $mask = chr((0xFF << (8 - $remainder)) & 0xFF);
        return (substr($rangeBin, $bytes, 1) & $mask) === (substr($ipBin, $bytes, 1) & $mask);
    }
}

File: aspnet/src/ShowAd.AspNetCore/AccessPolicy/AccessPolicyEvaluator.cs
using System.Net;
using Microsoft.AspNetCore.Http;

namespace ShowAd.AspNetCore.AccessPolicy;

public sealed class CrawlerVerificationResult
{
    public bool Verified { get; init; }
    public string Reason { get; init; } = string.Empty;
    public string? Family { get; init; }
}

/// <summary>
/// Pure access-policy evaluator. UA alone never bypasses; verified-crawler
/// requires a UA family match plus IP/CF/rDNS evidence.
/// </summary>
public sealed class AccessPolicyEvaluator
{
    public AccessPolicyDecision Evaluate(HttpContext context, AccessPolicyOptions options)
    {
        var clientIp = ResolveClientIp(context, options.TrustedIpHeaders);
        var userAgent = context.Request.Headers.UserAgent.ToString() ?? string.Empty;

        var crawler = VerifyCrawler(clientIp, userAgent, options.Crawler, context);
        if (crawler.Verified)
        {
            return AccessPolicyDecision.Allow($"crawler:{crawler.Family ?? "unknown"}");
        }

        if (!string.IsNullOrEmpty(clientIp) && IpInCidrs(clientIp!, options.AllowCidrs))
        {
            return AccessPolicyDecision.Allow("cidr_allowlist");
        }

        if (options.BeforeProtect is not null)
        {
            var ctx = new AccessPolicyContext
            {
                ClientIp = clientIp,
                UserAgent = userAgent,
            };
            return options.BeforeProtect(context, ctx) ?? AccessPolicyDecision.Continue();
        }

        return AccessPolicyDecision.Continue();
    }

    public string? ResolveClientIp(HttpContext context, IList<string> trustedHeaders)
    {
        foreach (var header in trustedHeaders)
        {
            if (!context.Request.Headers.TryGetValue(header, out var values)) continue;
            var raw = values.ToString();
            if (string.IsNullOrEmpty(raw)) continue;
            var first = raw.Split(',')[0].Trim();
            if (!string.IsNullOrEmpty(first)) return first;
        }
        return context.Connection.RemoteIpAddress?.ToString();
    }

    public CrawlerVerificationResult VerifyCrawler(string? ip, string userAgent, CrawlerPolicy? policy, HttpContext? context = null)
    {
        if (policy is null || !policy.Enabled)
            return new CrawlerVerificationResult { Verified = false, Reason = "disabled" };

        var families = policy.Families ?? CrawlerPolicy.DefaultUserAgents.Keys.ToList();
        var family = MatchFamily(userAgent, families, policy.UserAgents);
        if (family is null)
            return new CrawlerVerificationResult { Verified = false, Reason = "no_family_match" };

        if (string.IsNullOrEmpty(ip))
            return new CrawlerVerificationResult { Verified = false, Reason = "missing_ip", Family = family };

        if (policy.AllowCloudflareVerifiedBot && context is not null)
        {
            var raw = context.Request.Headers["CF-Verified-Bot"].ToString();
            if (string.IsNullOrEmpty(raw))
                raw = context.Request.Headers["X-ProofMark-CF-Verified-Bot"].ToString();
            if (IsTruthy(raw))
                return new CrawlerVerificationResult { Verified = true, Reason = "cloudflare_verified_bot", Family = family };
        }

        if (policy.FamilyCidrs.TryGetValue(family, out var cidrs) && IpInCidrs(ip!, cidrs))
            return new CrawlerVerificationResult { Verified = true, Reason = "cidr_match", Family = family };

        if (policy.ReverseDnsVerifier is not null && policy.ReverseDnsVerifier(ip!, family))
            return new CrawlerVerificationResult { Verified = true, Reason = "reverse_dns_match", Family = family };

        return new CrawlerVerificationResult { Verified = false, Reason = "ip_not_verified", Family = family };
    }

    public bool IpInCidrs(string ip, IEnumerable<string> cidrs)
    {
        foreach (var cidr in cidrs)
        {
            if (IpMatchesCidr(ip, cidr)) return true;
        }
        return false;
    }

    private static string? MatchFamily(string userAgent, IEnumerable<string> families, IDictionary<string, IList<string>>? overrides)
    {
        if (string.IsNullOrEmpty(userAgent)) return null;
        var needle = userAgent.ToLowerInvariant();

        foreach (var family in families)
        {
            IEnumerable<string>? fragments = null;
            if (overrides is not null && overrides.TryGetValue(family, out var ov)) fragments = ov;
            else if (CrawlerPolicy.DefaultUserAgents.TryGetValue(family, out var def)) fragments = def;
            if (fragments is null) continue;

            foreach (var frag in fragments)
            {
                if (string.IsNullOrEmpty(frag)) continue;
                if (needle.Contains(frag.ToLowerInvariant(), StringComparison.Ordinal))
                    return family;
            }
        }
        return null;
    }

    private static bool IsTruthy(string? value)
    {
        if (string.IsNullOrEmpty(value)) return false;
        var v = value.Trim().ToLowerInvariant();
        return v is "1" or "true" or "yes" or "on";
    }

    public static bool IpMatchesCidr(string ip, string cidr)
    {
        if (!IPAddress.TryParse(ip, out var ipAddr)) return false;

        var slash = cidr.IndexOf('/');
        if (slash < 0)
            return IPAddress.TryParse(cidr, out var single) && single.GetAddressBytes().AsSpan().SequenceEqual(ipAddr.GetAddressBytes());

        var rangeStr = cidr.Substring(0, slash);
        var bitsStr = cidr.Substring(slash + 1);

        if (!IPAddress.TryParse(rangeStr, out var rangeAddr)) return false;
        if (!int.TryParse(bitsStr, out var bits)) return false;

        var ipBytes = ipAddr.GetAddressBytes();
        var rangeBytes = rangeAddr.GetAddressBytes();
        if (ipBytes.Length != rangeBytes.Length) return false;

        var maxBits = ipBytes.Length * 8;
        if (bits < 0 || bits > maxBits) return false;

        var fullBytes = bits / 8;
        var remainder = bits % 8;

        for (var i = 0; i < fullBytes; i++)
        {
            if (ipBytes[i] != rangeBytes[i]) return false;
        }

        if (remainder == 0) return true;

        var mask = (byte)(0xFF << (8 - remainder));
        return (ipBytes[fullBytes] & mask) == (rangeBytes[fullBytes] & mask);
    }
}
